File: development/python/nlvd_test.py
import os.path as osp
import os


class NLVDTest:
    # dataset is a nlvd_dataset object created
    def __init__(self, dataset, test_dir, level_id):
        if not osp.exists(test_dir):
            os.makedirs(test_dir)
        self.dataset = dataset
        self.test_dir = test_dir
        self.level_id = level_id
        self.is_finished = False
        self.f = open(osp.join(self.test_dir, "level_%d.txt" % level_id), 'w')

    def text_ids(self, image_id):
        test_text_ids = self.dataset.test_text_ids(image_id, self.level_id)
        return test_text_ids

    # a function to write test result to a file in proper format
    # boxes_and_scores should be a map 
    # {phrase_id1: [[y1, x1, y2, x2, score1], [y1, x1, y2, x2, score2], ...]}
    # each phrase id maps with a list of detection results (bounding box coordinates with score)
    # all coordinates are 0-based
    def set_results(self, image_id, text_ids, boxes_and_scores):
        self.f.write(str(image_id) + ":")
        for t_id in text_ids:
            self.f.write("\n\t%s:" % t_id)
            # output the region informations
            regions = boxes_and_scores[t_id]
            for idx, box in enumerate(regions):
                # write in order [y1, x1, y2, x2]
                self.f.write(
                    " [%d, %d, %d, %d, %.6f]" %
                    (box[0] + 1, box[1] + 1, box[2] + 1, box[3] + 1, box[4])
                )
        self.f.write("\n")

    def finish(self):
        self.f.close()
        self.is_finished = True


nlvd_test = NLVDTest

File: development/matlab/nlvd_test.m
classdef nlvd_test < handle
    
    properties (GetAccess=public, SetAccess=protected)
        dataset
        test_dir
        level_id
        is_finished
        fid
    end

    methods (Access=public)
        function obj = nlvd_test(dataset, test_dir, level_id)
            obj.dataset = dataset;
            obj.test_dir = test_dir;
            obj.level_id = level_id;
            if ~exist(obj.test_dir, 'file')
                mkdir(obj.test_dir)
            end
            obj.is_finished = false;
            % create the text files
            obj.fid = fopen(sprintf('%s/level_%d.txt', test_dir, level_id), 'w');
        end
        
        function test_text_ids = text_ids(obj, image_id)
            test_text_ids = obj.dataset.test_text_ids( ...
                image_id, obj.level_id);
        end

        % boxes_and_scores is an cell array, each cell conrresponding to one text_id.
        % The array in the cell is [y1, x1, y2, x2, score] and the coordinate is 1-based
        function set_results(obj, image_id, text_ids, boxes_and_scores)
            fprintf(obj.fid, '%d:\n', image_id);
            for k = 1:numel(text_ids)
                tid = text_ids(k);
                bs = boxes_and_scores{k};
                fprintf(obj.fid,'\t%d:', tid);
                for j=1:numel(boxes_and_scores)
                    fprintf(obj.fid, ' [%d, %d, %d, %d, %.6f]', bs(j,1), bs(j,2), bs(j,3), bs(j,4), bs(j,5));
                end
                fprintf(obj.fid, '\n');
            end

        end
        
        function finish(obj)
            % close the result files
            fclose(obj.fid);
            obj.is_finished = true;
        end
        
        function delete(obj)
            if ~obj.is_finished
                obj.finish()
            end
        end
        
    end

end
